### PR TITLE
Vect reasoning library

### PR DESCRIPTION
**Pull request by ohad**
When working on Frex I needed a whole bunch of lemmata to do with `Data.Vect`. I hope it will be useful for others.

### CODE REVIEW

File: libs/contrib/Data/Vect/Properties.idr [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+||| Extensionally equivalent functions tabulate to the same vector
+export
+tabulateExtensional
+  : {n : Nat} -> (f, g : Fin n -> a) -> (ext : (i : Fin n) -> f i = g i)
+ -> tabulate f = tabulate g
+tabulateExtensional {n = 0  } f g ext = Refl
+tabulateExtensional {n = S n} f g ext =
+  rewrite ext FZ in
+  cong (g FZ ::) (tabulateExtensional (f . FS) (g . FS) (\ i => ext $ FS i))
```
Comment (gallais) [inline libs/contrib/Data/Vect/Properties.idr]: why not a `cong2 (::)` here?

File: libs/contrib/Data/Vect/Properties/Index.idr [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+||| Recalling by a position of `x` does yield `x`
+export
+recallElemSpec : (pos : x `Elem` xs) -> recallElem pos = x
+recallElemSpec  Here         = Refl
+recallElemSpec (There later) = recallElemSpec later
+
+||| map-fusion property for vectors up to function extensionality
+export
+mapFusionVect : (f : b -> c) -> (g : a -> b) -> (xs : Vect n a)
```
Comment (gallais) [inline libs/contrib/Data/Vect/Properties/Index.idr:27]: I don't think we need the `Vect` suffix. You can always use a qualified
name if you think there is an ambiguity

File: libs/contrib/Data/Vect/Properties/Tabulate.idr [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+||| Extensionally equivalent functions tabulate to the same vector
+export
+tabulateExtensional
+  : {n : Nat} -> (f, g : Fin n -> a) -> (ext : (i : Fin n) -> f i = g i)
+ -> tabulate f = tabulate g
+tabulateExtensional {n = 0  } f g ext = Refl
+tabulateExtensional {n = S n} f g ext =
+  rewrite ext FZ in
+  cong (g FZ ::) (tabulateExtensional (f . FS) (g . FS) (\ i => ext $ FS i))
```
Comment (gallais) [inline libs/contrib/Data/Vect/Properties/Tabulate.idr:27]: Why not a `cong2 (::)` here?

**Comment by ohad** [comment]
I think this is ready. I have a few more lemmas, but I'm not using them for anything, so allegedly they're useless :D

**Review by gallais — APPROVED** [review]

**Pull request merged**